parsing/location.ml

```diff
@@ -130,78 +132,108 @@ let highlight_terminfo ppf lb locs =
+
+     foo the_
+     error bar
+
+   gets displayed as:
+
+     ....the_
+     error....
+
+   If [locs] is empty then this function is a no-op.
+*)
+let highlight_dumb ~print_chars ppf lb locs =
+  if locs = [] then ()
+  else begin
+    let locs = List.map (fun loc ->
+      (loc.loc_start.pos_cnum, loc.loc_end.pos_cnum)
+    ) locs in
+    (* Helper to check if a given position is to be highlighted *)
+    let is_highlighted pos =
+      List.exists (fun (s, e) -> s <= pos && pos < e) locs in
+    (* Char 0 is at offset -lb.lex_abs_pos in lb.lex_buffer. *)
+    let pos0 = -lb.lex_abs_pos in
+    (* Helper to read a char in the buffer *)
+    let read_char pos = Bytes.get lb.lex_buffer (pos + pos0) in
+    (* Do nothing if the buffer does not contain the whole phrase. *)
+    if pos0 < 0 then raise Exit;
+    let end_pos = lb.lex_buffer_len - pos0 - 1 in
+    (* Leftmost starting position of all locations *)
+    let leftmost_start, _ =
+      List.hd @@ List.sort (fun (s, _) (s', _) -> compare s s') locs in
+    (* Rightmost ending position of all locations *)
+    let _, rightmost_end =
+      List.hd @@ List.sort (fun (_, e) (_, e') -> compare e' e) locs in
+    (* Determine line numbers for the start and end points *)
+    let line_start, line_end =
+      let line_start = ref 0 and line_end = ref 0 in
+      for pos = 0 to end_pos do
+        if read_char pos = '\n' then begin
+          if leftmost_start > pos then incr line_start;
+          if rightmost_end  > pos then incr line_end;
+        end
+      done;
+      !line_start, !line_end
+    in
+    Format.fprintf ppf "@[<v>";
+    (* Print character location (useful for Emacs) *)
+    if print_chars then begin
+      Format.fprintf ppf "Characters %i-%i:@,  @[<v>"
+        leftmost_start rightmost_end
+    end;
+    (* Print the input, highlighting the locations *)
+    let line = ref 0 in
+    let pos_at_bol = ref 0 in
+    for pos = 0 to end_pos do
+      match read_char pos with
+      | '\n' ->
+        if !line = line_start && !line = line_end then begin
+          (* single-line error: underline the locations *)
+          Format.fprintf ppf "@,";
+          for i = !pos_at_bol to rightmost_end - 1 do
+            if is_highlighted i then Format.pp_print_char ppf '^'
+            else Format.pp_print_char ppf ' '
+          done;
+        end;
+        if !line >= line_start && !line <= line_end then begin
+          Format.fprintf ppf "@,";
+        end;
+        incr line;
+        pos_at_bol := pos + 1
+      | '\r' -> () (* discard *)
+      | c ->
+        if !line = line_start && !line = line_end then
+          (* single-line error: print the whole line *)
           Format.pp_print_char ppf c
-        else
-          Format.pp_print_char ppf '.'
-      else if !line > !line_start && !line < !line_end then
-        (* intermediate line of multiline loc: print whole line *)
-        Format.pp_print_char ppf c
-  done;
-  Format.fprintf ppf "@]"
+        else begin
```

Comment:
I think the code would be more readable with just `else if` here, and a `begin..end` after the `then` instead.